### DISCUSSION TITLE
Sending a message through rabbit

### DISCUSSION THREAD

**Discussion by cristianleonie-geos** [celery/celery#9124] [Q&A] [answered]
## PROBLEM
I would like to send an email at X time (possibly a day later) given that a model precondition on django is satisfied. To do this I would like to use celery, I have seen that I can do it using apply_asinc, but it is not recommended for a delay greater than minutes (especially with RabbitMQ), the other solution would be to run continuosly a process CELERY_BEAT, but I don't like this option as it doesn't really suit my use case. 

I could also use a task scheduler or a cronjob, but since the system is already setup to work with Celery and Rabbit I would like to use them

## PROPOSED SOLUTION
As a solution I would like to schedule the task using a triggering message from RabbitMQ, since using it I can schedule message delivery for an indefinite amount of time.

I have celery setted up with geonode, I've created a task in my django server and I'm able to call it programmatically using Celery.

```
@shared_task(
    name="service.tasks.check_model_state_and_send_email",
    queue="service",
    ignore_result=False,
)
def check_model_state_and_send_email(model_id):
...

```
I can call it like this and it works
check_model_state_and_send_email.delay(12)
`
Celery is using rabbitmq for the queues.

I am able to read from the queues and send messages to the queues, the message generated by calling the delay function looks like this:
`[
    [
        12
    ],
    {},
    {
        "callbacks": null,
        "errbacks": null,
        "chain": null,
        "chord": null
    }
]`

## SETUP

GEONODE_EXCHANGE = Exchange("default", type="topic", durable=True)
queue = Queue("service", GEONODE_EXCHANGE, routing_key="service", priority=0)

This is my triggering message sender
```
connection = pika.BlockingConnection(host)
message = ...
channel = connection.channel()
channel.queue_declare(queue='service.tasks.check_model_state_and_send_email', passive=True)
channel.basic_publish(exchange='default',
                      routing_key='service',
                      body=json.dumps(message))

```
## PROBLEM
Despite the message being consumed by Celery, it does not trigger the task, I have tried different formats for the message, from simply copying the content of "correct" messages, to generating ones suggested online

Not only it doesn't produce the task, but it doesn't generate any logs. that I can see when the task is executed correctly or it fails

**Comment by passaadellism** [comment, marked as answer]
i had the same problem and i fixed it:
`

RABBITMQ_HOST = 'localhost'  
QUEUE_NAME = 'celery'
connection = pika.BlockingConnection(pika.ConnectionParameters(host=RABBITMQ_HOST, port=5672, virtual_host='/', credentials=pika.PlainCredentials('admin', 'admin')))
channel = connection.channel()

channel.queue_declare(queue=QUEUE_NAME, durable=True)

properties = pika.BasicProperties(
    content_type='application/json',
    content_encoding='utf-8',
    delivery_mode=2, 
    headers={
        'argsrepr': "(1, '2', True)",
        'kwargsrepr': "{'k': {'text': 'test'}, 'j': [1, 2, 3]}",
        'id': '80a207b8-2c75-4a13-a321-bfe4288d149f',
        'root_id': '80e207a9-2c75-4213-a321-bfe4288d149f',
        'sender': 'pika',
        'origin': 'genXXX@XXX',
        'task': 'tasks.add'
    },
)

message_body = json.dumps([
    [1, "2", True],
    {"k": {"text": "test"}, "j": [1, 2, 3]},
    {"callbacks": None, "errbacks": None, "chain": None, "chord": None}
])

channel.basic_publish(
    exchange='',
    routing_key=QUEUE_NAME,
    body=message_body,
    properties=properties
)

print(" [x] Sent message to RabbitMQ")
connection.close()
`

follow this and you will be fine.

Also doing this
@app.task
def add(x,y,z,k={},j=2):
    pass
    
 the task name for the above will be tasks.add so follow that convention
